luks_device.py: Allow manipulate LUKS containers with label or UUID

mulatinho opened the pull request:
Per default, on newer versions, cryptsetup creates LUKS containers with VERSION 2 format, but with this change it allow user create containers with `LABEL` identifier as requested on #58973 expliciting `LUKS2` format. It also allow manipulation of LUKS containers with `UUID`

- Allow create a LUKS2 container format with label support
- Allow manipulate (open, close, modify) an LUKS container based on
  both label (LUKS2 format) or UUID (V1 ans V2) instead of using devices only.

This was tested on Fedora Core 30 server and a Slackware 14.2 server.

Fixes: #58973

How to test these new features
```
---
- name: test issue 56872
  gather_facts: no
  hosts: freeipa
  remote_user: root
  tasks:
    - name: create images
      command: dd if=/dev/urandom of=/dev/fedora/test200M bs=100M count=2

    - name: create LUKS container
      luks_device:
        device: "/dev/fedora/test200M"
        state: "present"
        keyfile: "/vault/keyfile"
        label: "personalLabelName"

    - name: open LUKS container
      luks_device:
        state: "opened"
        name: "cryptDiskVol"
        keyfile: "/vault/keyfile"
        label: "personalLabelName"

    - name: close LUKS container
      luks_device:
        state: "closed"
        uuid: ae97ed88-b6eb-40e1-a32e-c2f08d8f0de2
```

Files changed:
- lib/ansible/modules/crypto/luks_device.py

mulatinho committed:
luks_device.py: Allow manipulate LUKS containers with label or UUID

- Allow create a LUKS2 container format with label support
- Allow manipulate (open, close, modify) an LUKS container based on
  both label (LUKS2 format) or UUID instead of using devices only.

Fixes: #58973
Signed-off-by: Alexandre Mulatinho <alex@mulatinho.net>

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
         default: no
+    label:
+        description:
+            - "This option allow the user creates a LUKS2 format container with
```
```suggestion
            - "This option allow the user to create a LUKS2 format container with
```

mulatinho replied:
Fixed, thanks! ;-)

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
+    label:
+        description:
+            - "This option allow the user creates a LUKS2 format container with
+              label support, then late it could be used to identify container
```
```suggestion
              label support, respectively to identifiy the container by label on later usages.
```

mulatinho replied:
Fixed, thanks! ;-)

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -251,15 +317,22 @@ def is_luks(self, device):
+        if label is not None:
+            # create luks container v2 with label
+            result = self._run_command([self._cryptsetup_bin, 'luksFormat',
+                                        '--type', 'luks2', '--label', label,
```
This will collide with #61522. Also, you should use a similar approach as in there, i.e. don't do an `if` with two copies of `self._run_command()`, but collect the options in a list and then finally call `self._run_command()`.

mulatinho replied:
Fixed, thanks! ;-)

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -251,15 +317,22 @@ def is_luks(self, device):
 
     def run_luks_open(self, device, keyfile, name):
         result = self._run_command([self._cryptsetup_bin, '--key-file', keyfile,
-                                    'open', '--type', 'luks', device, name])
```
Is there a reason why you want to remove this? Does this improve compatibility in any way, or is required for LUKS2 support?

mulatinho replied:
Its required for LUKS2 support, if pass `--type luks` it will make LUKS1 container on some versions.

felixfontein replied:
According to the man page, `luks` (the default value) opens both LUKS1 and LUKS2 containers; see here: http://man7.org/linux/man-pages/man8/cryptsetup.8.html#BASIC_COMMANDS

Do you know of a system where `--type luks` will not open LUKS2 containers? This is about opening, not creating.

mulatinho replied:
You are right sir, my bad!

felixfontein replied:
No problem. In fact, removing it shouldn't hurt, since `luks` is the default. But I guess it's better to keep it, just in case somewhen in the future, the default changes.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -336,14 +409,36 @@ class ConditionsHandler(Handler):
+            device = self._crypthandler.get_container_device_by_name(name)
+
+        if device is None:
+            error_msg = "Can't find the properly device based on label or " \
```
```suggestion
            error_msg = "Can't find the device based on label, UUID or name. " \
```

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -336,14 +409,36 @@ class ConditionsHandler(Handler):
+
+        if device is None:
+            error_msg = "Can't find the properly device based on label or " \
+                        "UUID, please review your playbook and check if the" \
```
```suggestion
                        "Please check whether the provided parameters are correct."
```

mulatinho replied:
Fixed, thanks! :)

mulatinho replied:
Done! Thanks ;D

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -336,14 +409,36 @@ class ConditionsHandler(Handler):
+        if device is None:
+            error_msg = "Can't find the properly device based on label or " \
+                        "UUID, please review your playbook and check if the" \
+                        " label/UUID informed was correct on remote"
```
```suggestion
```

mulatinho replied:
Done!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
+              label support, then late it could be used to identify container
+              instead of inform the device"
+        type: str
+        version_added: "2.9"
```
```suggestion
        version_added: "2.10"
```
(same below)

mulatinho replied:
Done!

felixfontein commented:
You'll also need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) (`minor_changes`).

mulatinho committed:
test_luks_device.py: organizing tests to support labels

- Add label on some tests and fix errors reported by Shippable

Signed-off-by: Alexandre Mulatinho <alex@mulatinho.net>

mulatinho committed:
luks_device.py: adjusting versions and messages

- Modifying version_added from 2.9 to 2.10
- Fixing some messages
- Created a changelog fragment
- Moving blkid from scope

Fixes #58973
Signed-off-by: Alexandre Mulatinho <alex@mulatinho.net>

mulatinho commented:
ready_for_review

felixfontein commented on changelogs/fragments/58973_luks_device-add-label-and-uuid-support.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- luks_device - added support to use uuid and labels with LUKS2 containers
```
```suggestion
- luks_device - added support to use UUIDs, and labels with LUKS2 containers
```
UUIDs also work with LUKS1 containers

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
+        version_added: "2.10"
+    uuid:
+        description:
+            - "With this option user can inform the LUKS container by UUID
```
```suggestion
            - "With this option user can identify the LUKS container by UUID.
```

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
+        description:
+            - "This option allow the user to create a LUKS2 format container
+              with label support, respectively to identifiy the container by
+              label on later usages."
```
```suggestion
              label on later usages."
            - "Will only be used on container creation, or when I(device) is not specified."
```

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -95,11 +95,26 @@
+    uuid:
+        description:
+            - "With this option user can inform the LUKS container by UUID
+              instead of pass the raw device"
```
```suggestion
            - "Will only be used when I(device) and I(label) are not specified."
```

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -153,6 +168,26 @@
+    keyfile: "/vault/keyfile"
+    label: personalLabelName
+
+- name: open the LUKS container based on label or UUID without device; name it "mycrypt"
```
```suggestion
- name: open the LUKS container based on label without device; name it "mycrypt"
```
You don't specify `uuid` here.

mulatinho replied:
You are correct, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -153,6 +168,26 @@
+    name: "mycrypt"
+    keyfile: "/vault/keyfile"
+
+- name: close container based on UUID or Label
```
```suggestion
- name: close container based on UUID
```

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -188,10 +223,41 @@ class Handler(object):
+        if uuid is None:
+            return None
+
+        query = "%s=%s" % ("UUID", uuid)
```
```suggestion
        query = '%s=%s' % ('UUID', uuid)
```
Since single quotes are used everywhere else.

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -188,10 +223,41 @@ class Handler(object):
+        if label is None:
+            return None
+
+        query = "%s=%s" % ("LABEL", label)
```
Same here.

mulatinho replied:
Done, thanks!

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -251,15 +317,21 @@ def is_luks(self, device):
+        command = [self._cryptsetup_bin, 'luksFormat', '-q', device, keyfile]
+        if label is not None:
+            # create luks container v2 with label
+            command = [self._cryptsetup_bin, 'luksFormat', '--type', 'luks2',
```
Instead of replacing `command`, why not extend it? This way it is very hard to extend this code.

felixfontein replied:
Anyway, it's probably best to wait until #61603 is merged and then change this.

mulatinho replied:
I took your suggestion, thanks!

felixfontein commented:
I've merged #61522, feel free to rebase any when you want :)

mulatinho commented:
> I've merged #61522, feel free to rebase any when you want :)

Cool! Thanks for the info, I will do soonly. :D

mulatinho commented:
ready_for_review

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -100,11 +100,29 @@
+    label:
+        description:
+            - "This option allow the user to create a LUKS2 format container
+              with label support, respectively to identifiy the container by
```
```suggestion
              with label support, respectively to identify the container by
```

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -100,11 +100,29 @@
+        description:
+            - "With this option user can identify the LUKS container by UUID."
+            - "Will only be used when I(device) and I(label) are not specified.
+              not specified."
```
```suggestion
```

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -193,10 +231,41 @@ class Handler(object):
+            return None
+
+        query = '%s=%s' % ('UUID', uuid)
+        result = self._run_command([self._blkid_bin, '-l', '-t', query])
```
Fun fact: when I run this as my user, it returns nothing. When I run it as root, it works. On the other hand, `blkid --uuid <UUID>` also works when I run it as my user.

So why not use
```suggestion
        result = self._run_command([self._blkid_bin, '--uuid', query])
```
instead? That seems to be simpler. Also, you don't need any postprocessing, since it returns only the device name and nothing else.

mulatinho replied:
Nice, much better sir, thanks ;-D

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -193,10 +231,41 @@ class Handler(object):
+            return None
+
+        query = '%s=%s' % ('LABEL', label)
+        result = self._run_command([self._blkid_bin, '-l', '-t', query])
```
Probably the same with `--label` instead of `--uuid` here.

mulatinho replied:
Nice, much better sir, thanks ;-D

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -193,10 +231,41 @@ class Handler(object):
     def __init__(self, module):
         self._module = module
         self._lsblk_bin = self._module.get_bin_path('lsblk', True)
+        self._blkid_bin = self._module.get_bin_path('blkid', True)
```
`blkid` is not always installed (for example Debian 9 and 10 don't necessarily include it). How about not requiring it (i.e. use `False`), and checking whether `self._blkid_bin` contains a result when you actually need it, and if then it doesn't, fail the module? Then `blkid` is not a requirement to run the module at all (right now it is).

felixfontein replied:
You can also move this line into `get_device_by_uuid()` and `get_device_by_label()`; then Ansible will fail for you if it cannot find the command, and it will only search for it when we actually need it.

mulatinho replied:
Hmmm, I choose the second option, thanks for the tip!

mulatinho commented:
ready_for_review

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -100,11 +100,28 @@
     - "cryptsetup"
     - "wipefs"
     - "lsblk"
+    - "blkid"
```
```suggestion
    - "blkid (when I(label) or I(uuid) options are used)"
```

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -100,11 +100,28 @@
 
 requirements:
     - "cryptsetup"
     - "wipefs"
```
While we're at it:
```suggestion
    - "wipefs (when I(state) is C(absent))"
```

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -85,52 +97,62 @@ def run_command_check(self, command):
     ("dummy", "key", "closed", "name", None, False),
     ("dummy", "key", "opened", "name", None, True),
-    (None, "key", "opened", "name", None, False),
+    (None, "key", "opened", "name", None, True),
```
Why did this value change?

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -85,52 +97,62 @@ def run_command_check(self, command):
+        conditions = luks_device.ConditionsHandler(module, crypt)
+        assert conditions.luks_create() == expected
+    except ValueError:
+        assert expected == "exception" or expected is False
```
```suggestion
        assert expected == "exception"
```
If `expected` is `False`, an exception should not be accepted.

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -145,8 +167,11 @@ def test_luks_remove(device, state, is_luks, expected, monkeypatch):
+        conditions = luks_device.ConditionsHandler(module, crypt)
+        assert conditions.luks_remove() == expected
+    except ValueError:
+        assert expected == "exception" or expected is False
```
```suggestion
        assert expected == "exception"
```

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -190,39 +221,53 @@ def test_luks_close(device, dev_by_name, name, name_by_dev, state,
+        conditions = luks_device.ConditionsHandler(module, crypt)
+        assert conditions.luks_close() == expected
+    except ValueError:
+        assert expected == "exception" or expected is False
```
```suggestion
        assert expected == "exception"
```

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -190,39 +221,53 @@ def test_luks_close(device, dev_by_name, name, name_by_dev, state,
         assert conditions.luks_add_key() == expected
     except ValueError:
-        assert expected == "exception"
+        assert expected == "exception" or expected is False
```
```suggestion
        assert expected == "exception"
```

felixfontein commented on test/units/modules/crypto/test_luks_device.py:
```diff
@@ -190,39 +221,53 @@ def test_luks_close(device, dev_by_name, name, name_by_dev, state,
         assert conditions.luks_remove_key() == expected
     except ValueError:
-        assert expected == "exception"
+        assert expected == "exception" or expected is False
```
```suggestion
        assert expected == "exception"
```

mulatinho replied:
I was making some tests and forget to clean this line, these tests are based on a lot of "fake mock results"

> It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

oh man ;/ sorry, let me fix that

felixfontein reviewed (commented):
The module code looks good (except two minor details, only one of them related to this PR). I have some questions about the unit tests, though.

mulatinho commented:
ready_for_review

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -100,11 +100,28 @@
 requirements:
     - "cryptsetup"
-    - "wipefs"
+    - "wipefs (when I(state) or C(absent))"
```
Almost ;)
```suggestion
    - "wipefs (when I(state) is C(absent))"
```

mulatinho replied:
LOL, laughing to not cry :P

felixfontein replied:
happens ;)

mulatinho commented:
ready_for_review

felixfontein commented:
It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

mulatinho commented:
> It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.

My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.

ready_for_review

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -197,6 +234,30 @@ def __init__(self, module):
+        uuid = self._module.params.get('uuid')
+        if uuid is None:
+            return None
+        result = self._run_command([self._blkid_bin, '--uid', uuid])
```
```suggestion
        result = self._run_command([self._blkid_bin, '--uuid', uuid])
```
According to `man blkid` on my system, it is `--uuid` and not `--uid`.

mulatinho replied:
> > > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> > 
> > 
> > My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

I always tought that create a lot of commits was not a good idea, but some of the rebases made here was not cool also, lesson is: dont write code with a late pregnant wife and with your kid of 2 wanting to play, wait for a calm day and write these on the right way (thanks God for all that). But its always the coder not the family behind him :D

sorry for made you waste your time, hope now, everything is OK.

felixfontein commented on lib/ansible/modules/crypto/luks_device.py:
```diff
@@ -197,6 +234,30 @@ def __init__(self, module):
+        ''' Returns the device that holds UUID passed by user
+        '''
+        self._blkid_bin = self._module.get_bin_path('blkid', True)
+        uuid = self._module.params.get('uuid')
```
The
```suggestion
        uuid = self._module.params['uuid']
```
you had here before is also fine. `AnsibleModule` makes sure that all options will be there; if the user didn't specify them, you'll get `None`.

mulatinho replied:
I got some errors without the .get method on `test_luks_device.py`, not on this place exactly. dont know why but they are there, well, I change this line and the another on label.

felixfontein commented:
> > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> 
> My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.

That's why it's best not to rebase, but to add new commits :)

mulatinho commented:
> > > It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> > 
> > 
> > My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

> 
> 
>         It looks like you pushed the wrong version, the current version undoes a lot of things from the previous pushes.
> 
>     My bad, two computers, a lot of rebases, I revised last versions and seems its OK now.
> 
> That's why it's best not to rebase, but to add new commits :)

I always tought that create a lot of commits was not a good idea, but some of the rebases made here was not cool also, lesson is: dont write code with a late pregnant wife and with your kid of 2 wanting to play, wait for a calm day and write these on the right way (thanks God for all that). But its always the coder not the family behind him :D

sorry for made you waste your time, hope now, everything is OK.

mulatinho commented:
ready_for_review

felixfontein reviewed (approved):
LGTM

felixfontein commented:
@mulatinho thank you very much for implementing this! :)
